test/terraform_aws_nextjs_serverless_test.go

```diff
@@ -0,0 +1,166 @@
+
+	// This is the cleanup part
+	defer test_structure.RunTestStage(t, "cleanupNextJSCloudFront", func() {
+		undeployTerraform(t, terraformDir, varFiles)
+	})
+
+	// This is the setup
+	test_structure.RunTestStage(t, "initializeNextJSCloudFront", func() {
+		deployTerraform(t, terraformDir, varFiles)
+	})
+
+	// To get the URL, we need to parse the terraform output
+	terraformOptions := terraform.WithDefaultRetryableErrors(t, &terraform.Options{
+		TerraformDir: terraformDir,
+		Reconfigure:  true,
+		VarFiles:     varFiles,
+	})
+	serverlessModule := terraform.OutputJson(t, terraformOptions, "next_serverless")
+	resultStat := gjson.Parse(fmt.Sprint(serverlessModule))
+	aliasURL := resultStat.Get("static-deploy.next_distribution.aliases.0").String()
+	if aliasURL == "" {
+		aliasURL = resultStat.Get("static-deploy.next_distribution.domain_name").String()
+	}
+
+	// Stage 1
+	test_structure.RunTestStage(t, "serverHealthCheck", func() {
+		hitTheFrontPage(t, aliasURL)
+	})
+
+	// Stage 2
+	test_structure.RunTestStage(t, "publicAssetsCheck", func() {
+		getIcons(t, aliasURL)
+	})
+
+	// Stage 3
+	// I could've checked the S3 buckets to see if these resources are uploaded there, too
+	// but assuming that terraform apply finished with success, I just need to check the build
+	// output of the NextJS app: required-server-files.json and build-manifest.json
+	// fileList := getRequiredServerFiles(aliasURL) //apparently these files do not exist in the static asset bucket
+	// fileList = append(fileList, getStaticAssets(aliasURL)...)
+	fileList := getStaticAssets(aliasURL)
+	fmt.Println(fileList)
+
+	test_structure.RunTestStage(t, "staticAssetsCheck", func() {
+		checkIfFilesExist(t, fileList, aliasURL)
+	})
+}
+
+func deployTerraform(t *testing.T, workingDir string, varfiles []string) {
+	terraformOptions := terraform.WithDefaultRetryableErrors(t, &terraform.Options{
+		TerraformDir: workingDir,
+		Reconfigure:  true,
+		VarFiles:     varfiles,
+	})
+	terraform.InitAndApply(t, terraformOptions)
+}
+
+func undeployTerraform(t *testing.T, workingDir string, varfiles []string) {
+	terraformOptions := terraform.WithDefaultRetryableErrors(t, &terraform.Options{
+		TerraformDir: workingDir,
+		Reconfigure:  true,
+		VarFiles:     varfiles,
+	})
+	terraform.Destroy(t, terraformOptions)
+}
+
+// the front page should contain next.js in the raw text and it should respond with 200
+func hitTheFrontPage(t *testing.T, aliasURL string) {
+	tlsConfig := tls.Config{}
+	code, body := http_helper.HttpGet(t, "https://"+aliasURL, &tlsConfig)
+	assert.Equal(t, code, 200)
+	assert.Contains(t, strings.ToLower(body), "next.js")
+}
+
+// hitting icons, which are public assets
+func getIcons(t *testing.T, aliasURL string) {
+	tlsConfig := tls.Config{}
+	code, _ := http_helper.HttpGet(t, "https://"+aliasURL+"/assets/vercel.svg", &tlsConfig)
+	assert.Equal(t, code, 200)
+	code, _ = http_helper.HttpGet(t, "https://"+aliasURL+"/assets/next.svg", &tlsConfig)
```

Comment:
Aren't we checking all assests dynamically anyways?

Reply:
No, these will be served from the public bucket.